ext/storage_sources/s3_store/aws_bucket_conn.cpp

```diff
@@ -0,0 +1,119 @@
+#include <aws/s3-crt/model/Object.h>
+#include <aws/s3-crt/model/PutObjectRequest.h>
+#include "aws_bucket_conn.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+/*
+ * list_buckets --
+ *     Return list of buckets from AWS account.
+ */
+bool
+aws_bucket_conn::list_buckets(std::vector<std::string> &buckets) const
+{
+    auto outcome = m_s3_crt_client.ListBuckets();
+    if (outcome.IsSuccess()) {
+        for (const auto& bucket : outcome.GetResult().GetBuckets())
+            buckets.push_back(bucket.GetName());
+        return true;
+    } else {
+        std::cerr << "Error in list_buckets: " << outcome.GetError().GetMessage() << std::endl
+                  << std::endl;
+        return false;
+    }
+}
+
+/*
+ * list_objects --
+ *     Return list of objects from S3 bucket.
+ */
+bool
+aws_bucket_conn::list_objects(
+  const Aws::String bucket_name, std::vector<Aws::S3Crt::Model::Object> &bucket_objects) const
+{
+    Aws::S3Crt::Model::ListObjectsRequest request;
+    request.WithBucket(bucket_name);
+    Aws::S3Crt::Model::ListObjectsOutcome outcomes = m_s3_crt_client.ListObjects(request);
+
+    if (outcomes.IsSuccess()) {
+        bucket_objects = outcomes.GetResult().GetContents();
+        return true;
+    } else {
+        std::cerr << "Error in list_buckets: " << outcomes.GetError().GetMessage()
+                  << std::endl;
+        return false;
+    }
+}
+
+/*
+ * put_object --
+ *     Put an object into S3 bucket.
+ */
+bool
+aws_bucket_conn::put_object(
+  const Aws::String &bucket_name, const Aws::String &object_key, const Aws::String &file_name) const
+{
+    Aws::S3Crt::Model::PutObjectRequest request;
+    request.SetBucket(bucket_name);
+    request.SetKey(object_key);
+
+    std::shared_ptr<Aws::IOStream> input_data = Aws::MakeShared<Aws::FStream>(
+      "s3-source", file_name.c_str(), std::ios_base::in | std::ios_base::binary);
+
+    if (!input_data->good()) {
+        std::cerr << "Error in put_object: Failed to open file'" << file_name << "'." << std::endl;
+        return false;
+    }
+
+    request.SetBody(input_data);
+
+    Aws::S3Crt::Model::PutObjectOutcome outcome = m_s3_crt_client.PutObject(request);
+
+    if (outcome.IsSuccess()) {
+        return true;
+    } else {
+        std::cerr << "Error in put_object: " << outcome.GetError().GetMessage() << std::endl;
+        return false;
+    }
+}
+
+/*
+ * delete_object --
+ *     Delete an object from S3 bucket.
+ */
+bool
+aws_bucket_conn::delete_object(const Aws::String &bucket_name, const Aws::String &object_key) const
+{
+    Aws::S3Crt::Model::DeleteObjectRequest request;
+    request.SetBucket(bucket_name);
+    request.SetKey(object_key);
+
+    Aws::S3Crt::Model::DeleteObjectOutcome outcome = m_s3_crt_client.DeleteObject(request);
+
+    if (outcome.IsSuccess()) {
+        return true;
+    } else {
+        std::cerr << "Error in delete_object: " << outcome.GetError().GetMessage() << std::endl;
```

Comment:
Do we need to consider how wiredtiger is implemented at this level?

Reply:
Not yet - we will do logging in a ticket in the next sprint.